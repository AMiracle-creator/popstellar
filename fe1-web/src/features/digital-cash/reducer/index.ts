export { default as digitalCashReducer } from './DigitalCashReducer';
export * from './DigitalCashReducer';
