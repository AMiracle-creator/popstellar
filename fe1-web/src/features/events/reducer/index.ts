export { default as eventsReducer } from './EventReducer';
export * from './EventReducer';
