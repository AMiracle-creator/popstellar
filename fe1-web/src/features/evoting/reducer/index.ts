export * from './ElectionKeyReducer';
export { default as electionKeyReducer } from './ElectionKeyReducer';
