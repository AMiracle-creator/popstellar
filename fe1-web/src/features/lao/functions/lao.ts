import { Channel, channelFromIds, Hash } from 'core/objects';
import { getStore } from 'core/redux';

import {
  makeLaoOrganizerBackendPublicKeySelector,
  selectCurrentLao,
  selectCurrentLaoId,
  selectIsLaoWitness,
  getLaoById as getLaoByIdFromState,
} from '../reducer';

/**
 * Returns the current lao and throws an error if there is none
 * @returns The current lao
 */
export const getCurrentLao = () => {
  const currentLao = selectCurrentLao(getStore().getState());

  if (!currentLao) {
    throw new Error('Error encountered while accessing storage : no currently opened LAO');
  }

  return currentLao;
};

/**
 * Returns the current lao and throws an error if there is none
 * @returns The current lao
 */
export const getLaoById = (laoId: string) => getLaoByIdFromState(laoId, getStore().getState());

/**
 * Returns the current lao id or undefined if there is none
 * @returns The current lao id
 */
export const getCurrentLaoId = () => selectCurrentLaoId(getStore().getState());

/**
 * Returns the organizer backend's public key for a given lao
 * @param laoId The lao id
 * @returns The organizer's backend public key for the given lao or undefined if it is not known
 */
export const getLaoOrganizerBackendPublicKey = (laoId: string) =>
  makeLaoOrganizerBackendPublicKeySelector(laoId)(getStore().getState());

/**
 * Returns whether the user is witness of the current lao
 */
export const isLaoWitness = () => selectIsLaoWitness(getStore().getState());

/**
 * Returns the lao organizer's public key
 * @param laoId the lao id
 */
export const getLaoOrganizer = (laoId: string) => getLaoById(laoId)?.organizer;

/**
 * Get a LAOs channel by its id
 * @param laoId The id of the lao whose channel should be returned
 * @returns The channel related to the passed lao id or undefined it the lao id is invalid
 */
export function getLaoChannel(laoId: string): Channel | undefined {
  try {
    return channelFromIds(new Hash(laoId));
  } catch (error) {
    console.error(`Cannot connect to LAO '${laoId}' as it is an invalid LAO ID`, error);
  }

  return undefined;
}
