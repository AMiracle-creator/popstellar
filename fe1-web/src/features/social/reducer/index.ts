export { default as socialReducer } from './SocialReducer';
export * from './SocialReducer';
