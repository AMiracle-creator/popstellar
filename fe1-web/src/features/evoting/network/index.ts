import { ActionType, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash } from 'core/objects';
import { dispatch, getStore } from 'core/redux';

import { EvotingConfiguration } from '../interface';
import { Election } from '../objects';
import { addElection, getElectionById, updateElection } from '../reducer';
import {
  handleCastVoteMessage,
  handleElectionEndMessage,
  handleElectionOpenMessage,
  handleElectionResultMessage,
  handleElectionSetupMessage,
} from './ElectionHandler';
import { CastVote, ElectionResult, EndElection, SetupElection } from './messages';
import { OpenElection } from './messages/OpenElection';

/**
 * Configures the network callbacks in a MessageRegistry.
 * @param configuration - An evoting config object
 */
export const configureNetwork = (configuration: EvotingConfiguration) => {
  // getElectionById bound to the global state
  const boundGetElectionById = (electionId: Hash | string) =>
    getElectionById(electionId, getStore().getState());

  const addElectionEvent = (laoId: Hash | string, election: Election) => {
    const electionState = election.toState();

    dispatch(addElection(electionState));
    dispatch(
      configuration.addEvent(laoId, {
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  const updateElectionEvent = (election: Election) => {
    const electionState = election.toState();

    dispatch(updateElection(electionState));
    dispatch(
      configuration.updateEvent({
        eventType: Election.EVENT_TYPE,
        id: electionState.id,
        start: electionState.start,
        end: electionState.start,
      }),
    );
  };

  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.SETUP,
    handleElectionSetupMessage(addElectionEvent),
    SetupElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.OPEN,
    handleElectionOpenMessage(boundGetElectionById, updateElectionEvent),
    OpenElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.CAST_VOTE,
    handleCastVoteMessage(boundGetElectionById, updateElectionEvent),
    CastVote.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.END,
    handleElectionEndMessage(boundGetElectionById, updateElectionEvent),
    EndElection.fromJson,
  );
  configuration.messageRegistry.add(
    ObjectType.ELECTION,
    ActionType.RESULT,
    handleElectionResultMessage(boundGetElectionById, updateElectionEvent),
    ElectionResult.fromJson,
  );
};
