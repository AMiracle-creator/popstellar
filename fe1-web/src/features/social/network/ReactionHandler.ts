import { ExtendedMessage, MessageRegistry } from 'core/network/jsonrpc/messages';
import { dispatch, getStore } from 'core/redux';
import { ActionType, ObjectType } from 'core/network/jsonrpc/messages/MessageData';
import { makeCurrentLao } from 'features/lao/reducer';

import { Reaction } from '../objects';
import { AddReaction } from './messages/reaction';
import { addReaction } from '../reducer';

/**
 * Handler for social media chirp's reactions
 */
const getCurrentLao = makeCurrentLao();

/**
 * Handles an addReaction message by storing the reaction sent.
 *
 * @param msg - The extended message for adding a reaction
 */
function handleAddReactionMessage(msg: ExtendedMessage): boolean {
  if (msg.messageData.object !== ObjectType.REACTION || msg.messageData.action !== ActionType.ADD) {
    console.warn('handleAddReaction was called to process an unsupported message');
    return false;
  }

  const makeErr = (err: string) => `reaction/add was not processed: ${err}`;

  const storeState = getStore().getState();
  const lao = getCurrentLao(storeState);
  if (!lao) {
    console.warn(makeErr('no Lao is currently active'));
    return false;
  }

  const messageId = msg.message_id;
  const { sender } = msg;
  const reactionMessage = msg.messageData as AddReaction;

  const reaction = new Reaction({
    id: messageId,
    sender: sender,
    codepoint: reactionMessage.reaction_codepoint,
    chirpId: reactionMessage.chirp_id,
    time: reactionMessage.timestamp,
  });

  dispatch(addReaction(lao.id, reaction.toState()));
  return true;
}

/**
 * Configures the ReactionHandler in a MessageRegistry.
 *
 * @param registry - The MessageRegistry where we want to add the mapping
 */
export function configure(registry: MessageRegistry) {
  registry.addHandler(ObjectType.REACTION, ActionType.ADD, handleAddReactionMessage);
}
