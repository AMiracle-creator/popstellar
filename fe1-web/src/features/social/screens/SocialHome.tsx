import React, { useCallback, useContext, useMemo, useState } from 'react';
import { ListRenderItemInfo, StyleSheet, View, ViewStyle } from 'react-native';
import { FlatList } from 'react-native-gesture-handler';
import { useToast } from 'react-native-toast-notifications';
import { useSelector } from 'react-redux';

import ScreenWrapper from 'core/components/ScreenWrapper';
import { List, Spacing } from 'core/styles';
import { FOUR_SECONDS } from 'resources/const';

import { ChirpCard, TextInputChirp } from '../components';
import { SocialMediaContext } from '../context';
import { SocialHooks } from '../hooks';
import { requestAddChirp } from '../network/SocialMessageApi';
import { Chirp } from '../objects';
import { makeChirpsList } from '../reducer';

/**
 * UI for the Social Media home screen component
 */

const styles = StyleSheet.create({
  userFeed: {
    flexDirection: 'column',
  } as ViewStyle,
  chirpList: {
    marginTop: Spacing.x1,
  } as ViewStyle,
});

const SocialHome = () => {
  const { currentUserPopTokenPublicKey } = useContext(SocialMediaContext);
  const [inputChirp, setInputChirp] = useState('');
  const toast = useToast();
  const laoId = SocialHooks.useCurrentLaoId();
  const isConnected = SocialHooks.useConnectedToLao();

  if (laoId === undefined) {
    throw new Error('Impossible to render Social Home, current lao id is undefined');
  }

  // The publish button is disabled in offline mode and when the user public key is not defined
  const publishDisabled = !isConnected || !currentUserPopTokenPublicKey;

  const publishChirp = () => {
    // button is disabled if publicKey is not set
    if (!currentUserPopTokenPublicKey) {
      return;
    }

    requestAddChirp(currentUserPopTokenPublicKey, inputChirp, laoId)
      .then(() => {
        setInputChirp('');
      })
      .catch((err) => {
        console.error('Failed to post chirp, error:', err);
        toast.show(`Failed to post chirp, error: ${err}`, {
          type: 'danger',
          placement: 'top',
          duration: FOUR_SECONDS,
        });
      });
  };

  const chirps = useMemo(() => makeChirpsList(laoId), [laoId]);
  const chirpList = useSelector(chirps);

  const renderChirp = useCallback(
    ({ item: chirp, index: i }: ListRenderItemInfo<Chirp>) => (
      <ChirpCard chirp={chirp} isFirstItem={i === 0} isLastItem={i === chirpList.length - 1} />
    ),
    [chirpList],
  );

  return (
    <ScreenWrapper>
      <View style={styles.userFeed}>
        <TextInputChirp
          testID="new_chirp"
          value={inputChirp}
          onChangeText={setInputChirp}
          onPress={publishChirp}
          disabled={publishDisabled}
          currentUserPublicKey={currentUserPopTokenPublicKey}
        />
        <View style={[List.container, styles.chirpList]}>
          <FlatList
            data={chirpList}
            renderItem={renderChirp}
            keyExtractor={(item) => item.id.toString()}
          />
        </View>
      </View>
    </ScreenWrapper>
  );
};

export default SocialHome;
