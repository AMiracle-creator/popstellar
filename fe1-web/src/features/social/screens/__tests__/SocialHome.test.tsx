import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { act } from 'react-test-renderer';
import { combineReducers, createStore } from 'redux';

import { mockKeyPair, mockLaoState } from '__tests__/utils';
import { laoReducer, setCurrentLao } from 'features/lao/reducer';
import { requestAddChirp } from 'features/social/network/SocialMessageApi';
import SocialReducer from 'features/social/reducer/SocialReducer';

import SocialHome from '../SocialHome';

jest.mock('features/social/network/SocialMessageApi', () => {
  const actual = jest.requireActual('features/social/network/SocialMessageApi');
  return {
    ...actual,
    requestAddChirp: jest.fn(() => Promise.resolve()),
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

const mockStore = createStore(combineReducers({ ...laoReducer, ...SocialReducer }));
mockStore.dispatch(setCurrentLao(mockLaoState));

describe('SocialHome', () => {
  it('renders correctly', () => {
    const { toJSON } = render(
      <Provider store={mockStore}>
        <SocialHome currentUserPublicKey={mockKeyPair.publicKey} />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('is possible to publish chirps', async () => {
    const { getByTestId } = render(
      <Provider store={mockStore}>
        <SocialHome currentUserPublicKey={mockKeyPair.publicKey} />
      </Provider>,
    );

    const mockText = 'some chirp text';

    fireEvent.changeText(getByTestId('new_chirp_input'), mockText);
    fireEvent.press(getByTestId('new_chirp_publish'));

    await waitFor(() => {
      expect(requestAddChirp).toHaveBeenCalledWith(mockKeyPair.publicKey, mockText);
      expect(requestAddChirp).toHaveBeenCalledTimes(1);
    });
  });
});
