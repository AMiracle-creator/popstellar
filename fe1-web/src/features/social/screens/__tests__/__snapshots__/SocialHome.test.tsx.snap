// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialHome renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "width": 600,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "flex-start",
        "marginTop": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 25,
          "lineHeight": 26,
          "marginHorizontal": 10,
          "textAlign": "center",
        }
      }
    >
      Home
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "column",
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "flexDirection": "row",
          "padding": 10,
          "width": 600,
        }
      }
    >
      <View
        style={
          Object {
            "width": 60,
          }
        }
      >
        <blockies
          bgColor="#ffffff"
          scale={5}
          seed="J9fBzJV70Jk5c-i3277Uq4CmeL4t53WDfUghaK0HpeM="
          size={8}
          spotColor="#008ec2"
        />
      </View>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "column",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          multiline={true}
          numberOfLines={5}
          onChangeText={[Function]}
          placeholder="Your chirp"
          rejectResponderTermination={true}
          selectTextOnFocus={true}
          style={
            Object {
              "alignContent": "flex-end",
              "borderColor": "gray",
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
              "width": 520,
            }
          }
          testID="new_chirp_input"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignSelf": "flex-end",
              "flexDirection": "row",
              "fontSize": 18,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 25,
                "lineHeight": 26,
                "marginHorizontal": 10,
                "textAlign": "center",
              }
            }
          >
            300
          </Text>
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="new_chirp_publish"
          >
            <View
              style={
                Array [
                  Object {},
                  Object {},
                ]
              }
            >
              <Text
                disabled={true}
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#cdcdcd",
                    },
                  ]
                }
              >
                Publish
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;
